sdk/nodejs/resource.ts

```diff
@@ -676,3 +676,160 @@ export class ComponentResource extends Resource {
 export const testingOptions = {
     isDryRun: false,
 };
+
+
+/**
+ * [mergeOptions] takes two ResourceOptions values and produces a new ResourceOptions with the
+ * respective properties of `opts2` merged over the same properties in `opts1`.  The original
+ * options objects will be unchanged.
+ *
+ * Conceptually property merging follows these basic rules:
+ *  1. if the property is a collection, the final value will be a collection containing the values
+ *     from each options object.
+ *  2. Simple scaler values from `opts2` (i.e. strings, numbers, bools) will replace the values of
+ *     `opts1`.
+ *  3. `opts2` can have properties explicitly provided with `null` or `undefined` as the value. If
+ *     explicitly provided, then that will be the final value in the result.
+ *  4. For the purposes of merging `dependsOn`, `provider` and `providers` are always treated as
+ *     collections, even if only a single value was provided.
+ */
```

Comment:
Should we add an implementation comment to ResourceOptions noting that any additions of new fields there should consider impact on this merging function?  Feels like there are many additions we might make there that could potentially break this in subtle ways.

Reply:
yup.  i will ad.d